lib/Default/MySqlDatabase.class.php

```diff
@@ -37,42 +48,42 @@ public function __construct() {
 	 * @param string $databaseName The name of the database to switch to
 	 */
 	public function switchDatabases(string $databaseName) {
-		self::$dbConn->select_db($databaseName);
-		self::$selectedDbName = $databaseName;
+		$this->dbConn->select_db($databaseName);
+		$this->selectedDbName = $databaseName;
 	}
 
 	/**
 	 * Switch back to the configured live database
 	 */
 	public function switchDatabaseToLive() {
-		$this->switchDatabases(self::$mysqlProperties->getDatabaseName());
+		$this->switchDatabases($this->mysqlProperties->getDatabaseName());
 	}
 
 	/**
 	 * Returns the size of the selected database in bytes.
 	 */
 	public function getDbBytes() {
-		$query = 'SELECT SUM(data_length + index_length) as db_bytes FROM information_schema.tables WHERE table_schema=' . $this->escapeString(self::$selectedDbName);
-		$result = self::$dbConn->query($query);
+		$query = 'SELECT SUM(data_length + index_length) as db_bytes FROM information_schema.tables WHERE table_schema=' . $this->escapeString($this->selectedDbName);
+		$result = $this->dbConn->query($query);
 		return (int)$result->fetch_assoc()['db_bytes'];
 	}
 
 	// This should not be needed except perhaps by persistent connections
 	public function close() {
-		if (self::$dbConn) {
-			self::$dbConn->close();
-			self::$dbConn = false;
+		if ($this->dbConn) {
+			$this->dbConn->close();
+			unset($this->dbConn);
```

Comment:
Try `getNewInstance`

Reply:
Are you sure that works? If I understand correctly, a new instance of `MySqlDatabase` will still be using the `mysqli` instance from the DI container, and thus the same connection, which will have been closed. Also, in this case, it won't necessarily be known where the next call to `getInstance` would be (it may be a place that, under most conditions, should use the existing `MySqlDatabase` instance, and thus can't be replaced with `getNewInstance`).

I realize the original design of `MySqlDatabase` is probably contributing to some of these problems (since it conflates the idea of connection vs. result, as one of its many flaws). It's not yet clear to me if additional refactoring is needed to solve this, or if there's a simple solution like the one you proposed.

Reply:
You're right I was in a hurry to reply on mobile and didn't think it through. I will look in the codebase and see where close is getting called and try to find an appropriate way to solve the issue.

I feel like views should not be accessing the database layer like that and having control over low level service routines like that. Also I can't think of a reason to close and reopen a connection to a database in the same page request. That being said I will look into a way to resolve this but yes I agree more refactoring is required. Eventually my goal is to remove accessing that class directly for any view objects

Reply:
Currently, this method is only used by the CLI programs, which need to close the connection between callbacks to avoid timeout errors (see 319c684eef). Like you, I can't think of a reason it would be needed for page requests.

Reply:
I also noticed that the `close()` method is used in `mysql_cleanup.php`. That `mysql_cleanup` function is referenced in a few script (they look like mostly Discord stuff).

I guess my question is, is there really a need to create new connections after the connection is closed? Is something in these CLI scripts opening and closing multiple times?

How necessary is it to have the ability to open another connection after it's been closed?

I guess we could have something equivalent to:

```php
public function query($query) {
	if (!$this->dbConn) {
		DiContainer::getContainer()
			->set(
				mysqli::class,
				DiContainer::make(mysqli::class));
		$this->dbConn = DiContainer::get(mysqli::class);
	}
	$this->dbResult = $this->dbConn->query($query);
}
```

Could be extracted to some sort of utility method somewhere, but I feel like maybe this is an edge case that doesn't really need to be resolved. I could be misunderstanding everything that's going on during the CLI calls, but if there really is ever a need for this edge case to be resolved we could circle back and do something about it then? Otherwise it looks like we'd be implementing something that wouldn't be used.

Reply:
It is currently essential to the proper functioning of the IRC and Discord bots that they be able to close the mysql connection between callbacks, because otherwise the connection times out and the bots no longer respond to user commands until they are restarted. If there is a way to solve this problem while keeping the mysql connection open, that would alleviate the need for a `close` method. I'm not a database expert, but I spent quite a long time debugging this back in 2017 and I couldn't find any other solution.

Reply:
I see, ok so they are long-lived processes these CLI guys. Hmm... Yeah well the `ping` command used to reconnect to the database, but looks like that is sorta a deprecated function now that PHP uses `mysqlnd`.

I think the only way to make this work is to implement some reconnect logic like above until one of the later refactors in which nothing uses `MySqlDatabase` directly.

Reply:
Perhaps here isn't the place, but could you give me some insight into your end-game with `MySqlDatabase`? What still uses this class? What do other parts of the code use instead? How would this redesign allow us to close and reopen mysql connections in an intuitive way? It sounds like you have it planned out in your head, so maybe you can share some spoilers. :)

This snippet above looks like it probably works if the next call is `$db->query(...)`, but obviously `query` isn't the only method that uses the `dbConn`. Maybe this stopgap needs to go into `getInstance` (and `getNewInstance`?) instead?

Reply:
About the location, I was just quickly spitballing something.

Ok so this might be long:

The end goal here is to separate concerns, encapsulate the code architecture into "domains of responsibility", and promote statelessness where possible.

It's a wide-sweeping thing that I'm _trying_ to do in functional (in the case of the `MySqlDatabase` class, temporary) pieces, and prevent breakage by preserving backwards compatibility. It started off with `MySqlDatabase`, since lots of things are touching it, but I plan on moving onto something like the "Account Domain" next.

One thing I'm trying to prevent is the `view` layer of the codebase from accessing the database directly. Instead, the views should use a `service` layer in order to do such things. Right now, IMHO, there is too much implementation detail about `MySqlDatabase` being leaked at too high of a level. Many views perform a query, check if there are results, iterate on those results, and other things. Those things could be handled in a service dedicated to that domain.

Want information about an account, or player? Use the account or player service. Want some voting information? Use the voting service.

You already have the `AbstractSmrAccount` class which is a combination of a `service`, `repository`, and a `model`/`entity`/`DAO`. I think we both agreed that this was pretty a "overloaded" class. It seemed the next logical step to work in since it doesn't seem like it gets updated all that often, so me separating that out into its constituent parts wouldn't interfere as much.

I'm also trying to promote "statelessness" in these classes. For instance, a big problem with `MySqlDatabase` class is that it has state for the `record` and the `result` of the queries. One of the longer-term goals is to remove that responsibility from this wrapper. We've talked, and you've mentioned, that there are conflation issues with this class. Not only is it a service, but it's also acting as a partial repository AND a stateful query result wrapper.

All of this leads back to a cleaner code structure, and one that is _easily_ testable whether it's an integration test for database repositories, or simple unit tests.

Here's a contrived example modern pattern I'm working to achieve.

```php
class AccountService {
    private AccountRepository $accountRepository;

    public function __construct(AccountRepository $accountRepository) {
        $this->accountRepository = $accountRepository;
    }

    public function findByEmail($email): Account {
        $account = $this->accountRepository->findByEmail($email);
        // ... Do some processing and return results
        return $account;
    }
}

class AccountRepository {
    private MySqlDatabase $mySqlDatabase;

    // In all actuality, I would want to remove many of the guts of MySqlDatabase at some point
    // So it can truly be a stateless class
    public function __construct(MySqlDatabase $mySqlDatabase) {
        $this->mySqlDatabase = $mySqlDatabase;
    }

    public function findByEmail($email): Account {
        // ... Write query, and return results, if any
    }
}

class Account {
    public $username;
    public $email;
}

// some_account_page.php
$account = $accountService->findByEmail("asd@asd.asd");
```

Let's note that the `findByEmail` sort of looks like a passthru, but the `AccountService` class would of course contain many more useful business methods than that. However I hope the comments inside the method show their responsibilities even though they share a common name.

I offered once, not sure if you remember or saw it, but we could create an architectural diagram of the current way things look, and an end goal. We can also get together and define these "domains of responsibility", and the way to approach migrating things over safely little-by-little.

Reply:
Some definitions:
**View layer**: The layer that the user sees. Should be responsible for displaying data retrieved from some service authority.
**Service layer**: This is the layer that performs all of the heavy lifting and business logic, and returns output to controller layers, and view layers. This layer is probably the most important, and should contain as much test coverage as possible.
**Repository layer**: This layer is responsible for querying a data store, and returning results

Reply:
Thanks for the insightful comments. It's useful to have a bit of concrete code to see where you're trying to get to.

Translating some of these terms to our existing infrastructure:

**view layer** is our `templates/` (though there's a bit too much database/object access in there, which is part of what https://github.com/smrealms/smr/issues/920 is trying to remedy).

**service layer** is our `engine/` (though again too much direct database access -- ideally it would just be using Smr library functions/classes).

And our `lib/` is pretty much a mishmash of everything else. :P

Reply:
Updated so that `getInstance()` and `getNewInstance()` will reconnect when needed